enterprise_access/apps/api/tasks.py

```diff
@@ -1,26 +1,94 @@
+from enterprise_access.apps.api_client.lms_client import LmsApiClient
+from enterprise_access.apps.subsidy_request.constants import (
+    ENTERPRISE_BRAZE_ALIAS_LABEL,
+    SUBSIDY_TYPE_CHANGE_DECLINATION,
+    SubsidyRequestStates,
+    SubsidyTypeChoices
+)
 from enterprise_access.apps.subsidy_request.models import CouponCodeRequest, LicenseRequest
 
+logger = logging.getLogger(__name__)
+
+
+def _aliased_recipient_object_from_email(user_email):
+    """
+    Returns a dictionary with a braze recipient object, including
+    a braze alias object.
+    """
+    return {
+        'attributes': {'email': user_email},
+        'user_alias': {
+            'alias_label': ENTERPRISE_BRAZE_ALIAS_LABEL,
+            'alias_name': user_email,
+        },
+    }
+
 
 @shared_task(base=LoggedTask)
-def delete_enterprise_subsidy_requests_task(enterprise_customer_uuid, subsidy_type):
+def decline_enterprise_subsidy_requests_task(enterprise_customer_uuid, subsidy_type, send_notification):
     """
-    Delete all subsidy requests of the given type for the enterprise customer.
+    Decline all subsidy requests of the given type for the enterprise customer.
     """
 
     if subsidy_type == SubsidyTypeChoices.COUPON:
-        CouponCodeRequest.objects.filter(
-            enterprise_customer_uuid=enterprise_customer_uuid
-        ).delete()
-
+        subsidy_requests = CouponCodeRequest.objects.filter(
+            enterprise_customer_uuid=enterprise_customer_uuid,
+            state__in=[
+                SubsidyRequestStates.REQUESTED,
+                SubsidyRequestStates.PENDING,
+                SubsidyRequestStates.ERROR
+            ],
+        )
     if subsidy_type == SubsidyTypeChoices.LICENSE:
-        LicenseRequest.objects.filter(
-            enterprise_customer_uuid=enterprise_customer_uuid
-        ).delete()
+        subsidy_requests = LicenseRequest.objects.filter(
+            enterprise_customer_uuid=enterprise_customer_uuid,
+            state__in=[
+                SubsidyRequestStates.REQUESTED,
+                SubsidyRequestStates.PENDING,
+                SubsidyRequestStates.ERROR
+            ],
+        )
+
+    # Why I don't used subsidy_requests.update() #
+    # When you run .update() on a queryset, you "lose" the objects, because by
+    # nature of them being updated in the DB (update runs raw SQL),
+    # they no longer are returned by the original
+    # queryset. To make sure we send out notifications for the exact objects we are
+    # declining here, I've opted to use a save() in a for-loop (which the django
+    # docs even recommend in some cases).
+    for subsidy_request in subsidy_requests:
+        logger.info(f'Declining subsidy {subsidy_request} because subsidy type changed on Configuration.')
+        subsidy_request.state = SubsidyRequestStates.DECLINED
+        subsidy_request.decline_reason = SUBSIDY_TYPE_CHANGE_DECLINATION
+        subsidy_request.save()
+
+    if send_notification:
```

Comment:
Sure, I can pull in that ticket and consider it part of this, though question:

I currently have this notification logic within the decline task (as you see), because it felt correct to keep the Config from pulling lots of DB objects from within the synchronous HTTP lifecycle, while still being able to only query the DB once for the SubsidyRequest objects. The AC of 5403 does specify the notifications being sent via their own task... how much do we care about that AC? As in, is the current set up (1 decline task with optional notifications being sent) reasonable?

Reply:
My initial thought is I care a lot about that AC. I would prefer for all these notifications to be sent asynchronously via the same code paths using a celery task. A bit much for a single request? Possibly... However I would like for us to reuse as much code as possible here and for the sake of troubleshooting I would prefer all mechanisms for sending these alerts follow the same process/code paths. 

I also like the idea of the asynchronous nature of this. The assignment should fail if braze is down. Emails should be kept in queue until braze comes back online. I think there is some additional work to get us to that last state, that said separating this out in to it's own celery task takes us the first step on this journey.

Reply:
So, okay, I will split the notifications into its own task